File: benchmark/legacy_demo/java/expenses/src/main/java/com/example/expenses/Budget.java
package com.example.expenses;

import java.util.List;

public final class Budget {

    private final double monthlyCap;
    private final double categoryCap;

    public Budget(double monthlyCap, double categoryCap) {
        this.monthlyCap = monthlyCap;
        this.categoryCap = categoryCap;
    }

    public double categoryTotal(List<Expense> expenses, String category) {
        double total = 0.0;
        for (Expense e : expenses) {
            if (e.category.equals(category)) {
                total += e.amount;
            }
        }
        return total;
    }

    public double total(List<Expense> expenses) {
        double t = 0.0;
        for (Expense e : expenses) {
            t += e.amount;
        }
        return t;
    }

    public String status(double spent) {
        if (spent > monthlyCap) {
            return "OVER_BUDGET";
        }
        if (spent > monthlyCap * 0.8) {
            return "WARNING";
        }
        return "OK";
    }

    public boolean allows(Expense e, List<Expense> all) {
        double categorySpent = categoryTotal(all, e.category);
        if (categorySpent + e.amount > categoryCap) {
            return false;
        }
        if (total(all) + e.amount > monthlyCap) {
            return false;
        }
        return e.amount > 0;
    }
}

File: benchmark/legacy_demo/python/expense/budget.py
"""Budget logic — the high-complexity module of the Python benchmark.

Nested branches, multiple conditions per branch, and a custom exception
give the test generator plenty of uncovered paths to target.
"""


class BudgetError(ValueError):
    pass


def category_total(expenses: list[dict], category: str) -> float:
    return sum(e["amount"] for e in expenses if e["category"] == category)


def total(expenses: list[dict]) -> float:
    return sum(e["amount"] for e in expenses)


def status(spent: float, cap: float) -> str:
    if cap <= 0:
        raise BudgetError("cap must be positive")
    if spent > cap:
        return "OVER_BUDGET"
    if spent > cap * 0.8:
        return "WARNING"
    return "OK"


def allows(expense: dict, expenses: list[dict], monthly_cap: float, category_cap: float) -> bool:
    if category_total(expenses, expense["category"]) + expense["amount"] > category_cap:
        return False
    if total(expenses) + expense["amount"] > monthly_cap:
        return False
    return expense["amount"] > 0


def remaining(expenses: list[dict], cap: float) -> float:
    return max(0.0, cap - total(expenses))
